README.rst
=============================
django-telegrambot
=============================

.. image:: https://badge.fury.io/py/django-telegrambot.png
    :target: https://badge.fury.io/py/django-telegrambot

.. image:: https://travis-ci.org/JungDev/django-telegrambot.png?branch=master
    :target: https://travis-ci.org/JungDev/django-telegrambot

.. image:: https://img.shields.io/badge/Donate-PayPal-green.svg
    :target: https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=LMXQVQ3YA2JJQ

.. image:: http://pepy.tech/badge/django-telegrambot
    :target: http://pepy.tech/count/django-telegrambot

A simple app to develop Telegram bots with Django

Documentation
-------------

The full documentation is at https://django-telegrambot.readthedocs.org.

If this project help you reduce time to develop, you can give me a cup of coffee :)

.. image:: https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif
    :target: https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=LMXQVQ3YA2JJQ


Changelog
------------
* **IMPORTANT ver 1.0.0** : If you upgrade from a previous version, you **MUST** change how to include ``django_telegrambot.urls`` and modify your ``settings.py``.


Quickstart
----------

Install django-telegrambot::

    pip install django-telegrambot

Configure your installation
---------------------------

Add ``django_telegrambot`` in ``INSTALLED_APPS`` ::

       #settings.py
       INSTALLED_APPS = (
           ...
           'django_telegrambot',
           ...
       )

And set your bots::

        #settings.py
        #Django Telegram Bot settings

        DJANGO_TELEGRAMBOT = {

            'MODE' : 'WEBHOOK', #(Optional [str]) # The default value is WEBHOOK,
                                # otherwise you may use 'POLLING'
                                # NB: if use polling you must provide to run
                                # a management command that starts a worker

            'WEBHOOK_SITE' : 'https://mywebsite.com',
            'WEBHOOK_PREFIX' : '/prefix', # (Optional[str]) # If this value is specified,
                                          # a prefix is added to webhook url

            #'WEBHOOK_CERTIFICATE' : 'cert.pem', # If your site use self-signed
                                 #certificate, must be set with location of your public key
                                 #certificate.(More info at https://core.telegram.org/bots/self-signed )

            'STRICT_INIT': True, # If set to True, the server will fail to start if some of the
                                 # apps contain telegrambot.py files that cannot be successfully
                                 # imported.

            'BOTS' : [
                {
                   'TOKEN': '123456:ABC-DEF1234ghIkl-zyx57W2v1u123ew11', #Your bot token.

                   #'ALLOWED_UPDATES':(Optional[list[str]]), # List the types of
                                   #updates you want your bot to receive. For example, specify
                                   #``["message", "edited_channel_post", "callback_query"]`` to
                                   #only receive updates of these types. See ``telegram.Update``
                                   #for a complete list of available update types.
                                   #Specify an empty list to receive all updates regardless of type
                                   #(default). If not specified, the previous setting will be used.
                                   #Please note that this parameter doesn't affect updates created
                                   #before the call to the setWebhook, so unwanted updates may be
                                   #received for a short period of time.

                   #'TIMEOUT':(Optional[int|float]), # If this value is specified,
                                   #use it as the read timeout from the server

                   #'WEBHOOK_MAX_CONNECTIONS':(Optional[int]), # Maximum allowed number of
                                   #simultaneous HTTPS connections to the webhook for update
                                   #delivery, 1-100. Defaults to 40. Use lower values to limit the
                                   #load on your bot's server, and higher values to increase your
                                   #bot's throughput.

                   # 'MESSAGEQUEUE_ENABLED':(Optinal[bool]), # Make this True if you want to use messagequeue

                   # 'MESSAGEQUEUE_ALL_BURST_LIMIT':(Optional[int]), # If not provided 29 is the default value

                   # 'MESSAGEQUEUE_ALL_TIME_LIMIT_MS':(Optional[int]), # If not provided 1024 is the default value

                   # 'MESSAGEQUEUE_REQUEST_CON_POOL_SIZE':(Optional[int]), # If not provided 8 is the default value

                   #'POLL_INTERVAL' : (Optional[float]), # Time to wait between polling updates from Telegram in
                                   #seconds. Default is 0.0

                   #'POLL_CLEAN':(Optional[bool]), # Whether to clean any pending updates on Telegram servers before
                                   #actually starting to poll. Default is False.

                   #'POLL_BOOTSTRAP_RETRIES':(Optional[int]), # Whether the bootstrapping phase of the `Updater`
                                   #will retry on failures on the Telegram server.
                                   #|   < 0 - retry indefinitely
                                   #|     0 - no retries (default)
                                   #|   > 0 - retry up to X times

                   #'POLL_READ_LATENCY':(Optional[float|int]), # Grace time in seconds for receiving the reply from
                                   #server. Will be added to the `timeout` value and used as the read timeout from
                                   #server (Default: 2).
                },
                #Other bots here with same structure.
            ],

        }



Include in your urls.py the ``django_telegrambot.urls`` (NB: If you upgrade from a previous version, you MUST change how to include ``django_telegrambot.urls``. Never set prefix here!)::

        #urls.py
        urlpatterns = [
            ...
            url(r'^', include('django_telegrambot.urls')),
            ...
        ]

Then use it in a project creating a module ``telegrambot.py`` in your app ::

        #myapp/telegrambot.py
        # Example code for telegrambot.py module
        from telegram.ext import CommandHandler, MessageHandler, Filters
        from django_telegrambot.apps import DjangoTelegramBot

        import logging
        logger = logging.getLogger(__name__)


        # Define a few command handlers. These usually take the two arguments bot and
        # update. Error handlers also receive the raised TelegramError object in error.
        def start(bot, update):
            bot.sendMessage(update.message.chat_id, text='Hi!')


        def help(bot, update):
            bot.sendMessage(update.message.chat_id, text='Help!')


        def echo(bot, update):
            bot.sendMessage(update.message.chat_id, text=update.message.text)


        def error(bot, update, error):
            logger.warn('Update "%s" caused error "%s"' % (update, error))


        def main():
            logger.info("Loading handlers for telegram bot")

            # Default dispatcher (this is related to the first bot in settings.DJANGO_TELEGRAMBOT['BOTS'])
            dp = DjangoTelegramBot.dispatcher
            # To get Dispatcher related to a specific bot
            # dp = DjangoTelegramBot.getDispatcher('BOT_n_token')     #get by bot token
            # dp = DjangoTelegramBot.getDispatcher('BOT_n_username')  #get by bot username

            # on different commands - answer in Telegram
            dp.add_handler(CommandHandler("start", start))
            dp.add_handler(CommandHandler("help", help))

            # on noncommand i.e message - echo the message on Telegram
            dp.add_handler(MessageHandler([Filters.text], echo))

            # log all errors
            dp.add_error_handler(error)



Features
--------

* Multiple bots
* Admin dashboard available at ``/admin/django-telegrambot``
* Polling mode by management command (an easy to way to run bot in local machine, not recommended in production!)

      ``(myenv) $ python manage.py botpolling --username=<username_bot>``
* Supporting messagequeues

Contributing
------------

Patches and bug reports are welcome, just please keep the style consistent with the original source.

Running Tests
--------------

Does the code actually work?

::

    source <YOURVIRTUALENV>/bin/activate
    (myenv) $ pip install -r requirements-test.txt
    (myenv) $ python runtests.py

Sample Application
------------------
There a sample application in `sampleproject` directory. Here is installation instructions:

1. Install requirements with command

        pip install -r requirements.txt
2. Copy file `local_settings.sample.py` as `local_settings.py` and edit your bot token

        cp sampleproject/local_settings.sample.py sampleproject/local_settings.py

        nano sampleproject/local_settings.py
3. Run Django migrations

        python manage.py migrate
4. Run server

        python manage.py runserver
5. If **WEBHOOK** Mode setted go to 8

6. If **POLLING** Mode setted, open in your browser http://localhost/

7. Open Django-Telegram Dashboard http://localhost/admin/django-telegrambot and follow instruction to run worker by management command `botpolling`. Then go to 10

8. To test webhook locally install `ngrok` application and run command

        ./ngrok http 8000
9. Change `WEBHOOK_SITE` and `ALLOWED_HOSTS` in local_settings.py file

10. Start a chat with your bot using telegram.me link avaible in **Django-Telegram Dashboard** at http://localhost/admin/django-telegrambot

Credits
---------
Required package:

*  `Python Telegram Bot`_

.. _`Python Telegram Bot`: https://github.com/python-telegram-bot/python-telegram-bot

Tools used in rendering this package:

*  Cookiecutter_

.. _Cookiecutter: https://github.com/audreyr/cookiecutter



HISTORY.rst
.. :changelog:

History
-------
1.1.0 (unreleased)
++++++++++++++++++
* Fix error on startup when using ``MESSAGEQUEUE_ENABLED`` [Nachatlb]

1.0.0 (2017-05-25)
++++++++++++++++++
* IMPORTANT: If you upgrade from a previous version, you MUST change how to include django_telegrambot.urls and settings.py.
* Added admin dashboard, available at /admin/django-telegrambot
* Added polling mode from management command (an easy to way to run bot in local machine, not recommended in production)
* More setting available
* Improved AppConfig
* Improved sample project

0.2.6 (2017-04-08)
++++++++++++++++++
* Improved module loading
* Added sample project

0.2.5 (2017-03-06)
++++++++++++++++++
* Fix compatibility with python-telegram-bot 5.1

0.2.4 (2016-10-04)
++++++++++++++++++
* Fix compatibility with Django 1.10

0.2.3 (2016-07-30)
++++++++++++++++++
* Fix default dispatcher and bot

0.2.2 (2016-07-27)
++++++++++++++++++
* Fix multi workers

0.2.1 (2016-07-24)
++++++++++++++++++
* Update for python-telegram-bot release v5.0

0.2.0 (2016-04-27)
++++++++++++++++++

* Update for python-telegram-bot release v4.0.1

0.1.8 (2016-03-22)
++++++++++++++++++

* Update for deprecation in python-telegram-bot release v3.4

0.1.5 (2016-01-28)
++++++++++++++++++

* Fix compatibility.

0.1.4 (2016-01-28)
++++++++++++++++++

* Fix compatibility.

0.1.3 (2016-01-28)
++++++++++++++++++

* Fix setting certificate.
* Add method DjangoTelegramBot.getBot(); get bot instance by token or username.

0.1.2 (2016-01-26)
++++++++++++++++++

* First release on PyPI.


django_telegrambot/apps.py
# coding=utf-8
# django_telegram_bot/apps.py
from django.apps import AppConfig
from django.apps import apps
from django.conf import settings
import importlib
import telegram
from django.utils.module_loading import module_has_submodule
from telegram.ext import Dispatcher
from telegram.ext import Updater
from telegram.error import InvalidToken, TelegramError
from telegram.utils.request import Request
from telegram.ext import messagequeue as mq
from .mqbot import MQBot
import os.path

import logging

logger = logging.getLogger(__name__)

TELEGRAM_BOT_MODULE_NAME = 'telegrambot'
WEBHOOK_MODE, POLLING_MODE = range(2)


class classproperty(property):
    def __get__(self, obj, objtype=None):
        return super(classproperty, self).__get__(objtype)

    def __set__(self, obj, value):
        super(classproperty, self).__set__(type(obj), value)

    def __delete__(self, obj):
        super(classproperty, self).__delete__(type(obj))


class DjangoTelegramBot(AppConfig):
    name = 'django_telegrambot'
    verbose_name = 'Django TelegramBot'
    ready_run = False
    bot_tokens = []
    bot_usernames = []
    dispatchers = []
    bots = []
    updaters = []
    __used_tokens = set()

    @classproperty
    def dispatcher(cls):
        # print("Getting value default dispatcher")
        cls.__used_tokens.add(cls.bot_tokens[0])
        return cls.dispatchers[0]

    @classproperty
    def updater(cls):
        # print("Getting value default updater")
        cls.__used_tokens.add(cls.bot_tokens[0])
        return cls.updaters[0]

    @classmethod
    def get_dispatcher(cls, bot_id=None, safe=True):
        if bot_id is None:
            cls.__used_tokens.add(cls.bot_tokens[0])
            return cls.dispatchers[0]
        else:
            try:
                index = cls.bot_tokens.index(bot_id)
            except ValueError:
                if not safe:
                    return None
                try:
                    index = cls.bot_usernames.index(bot_id)
                except ValueError:
                    return None
            cls.__used_tokens.add(cls.bot_tokens[index])
            return cls.dispatchers[index]

    @classmethod
    def getDispatcher(cls, bot_id=None, safe=True):
        return cls.get_dispatcher(bot_id, safe)

    @classmethod
    def get_bot(cls, bot_id=None, safe=True):
        if bot_id is None:
            if safe:
                return cls.bots[0]
            else:
                return None
        else:
            try:
                index = cls.bot_tokens.index(bot_id)
            except ValueError:
                if not safe:
                    return None
                try:
                    index = cls.bot_usernames.index(bot_id)
                except ValueError:
                    return None
            return cls.bots[index]

    @classmethod
    def getBot(cls, bot_id=None, safe=True):
        return cls.get_bot(bot_id, safe)

    @classmethod
    def get_updater(cls, bot_id=None, safe=True):
        if bot_id is None:
            return cls.updaters[0]
        else:
            try:
                index = cls.bot_tokens.index(bot_id)
            except ValueError:
                if not safe:
                    return None
                try:
                    index = cls.bot_usernames.index(bot_id)
                except ValueError:
                    return None
            return cls.updaters[index]

    @classmethod
    def getUpdater(cls, id=None, safe=True):
        return cls.get_updater(id, safe)

    def ready(self):
        if DjangoTelegramBot.ready_run:
            return
        DjangoTelegramBot.ready_run = True

        self.mode = WEBHOOK_MODE
        if settings.DJANGO_TELEGRAMBOT.get('MODE', 'WEBHOOK') == 'POLLING':
            self.mode = POLLING_MODE

        modes = ['WEBHOOK', 'POLLING']
        logger.info('Django Telegram Bot <{} mode>'.format(modes[self.mode]))

        bots_list = settings.DJANGO_TELEGRAMBOT.get('BOTS', [])

        if self.mode == WEBHOOK_MODE:
            webhook_site = settings.DJANGO_TELEGRAMBOT.get('WEBHOOK_SITE', None)
            if not webhook_site:
                logger.warn('Required TELEGRAM_WEBHOOK_SITE missing in settings')
                return
            if webhook_site.endswith("/"):
                webhook_site = webhook_site[:-1]

            webhook_base = settings.DJANGO_TELEGRAMBOT.get('WEBHOOK_PREFIX', '/')
            if webhook_base.startswith("/"):
                webhook_base = webhook_base[1:]
            if webhook_base.endswith("/"):
                webhook_base = webhook_base[:-1]

            cert = settings.DJANGO_TELEGRAMBOT.get('WEBHOOK_CERTIFICATE', None)
            certificate = None
            if cert and os.path.exists(cert):
                logger.info('WEBHOOK_CERTIFICATE found in {}'.format(cert))
                certificate = open(cert, 'rb')
            elif cert:
                logger.error('WEBHOOK_CERTIFICATE not found in {} '.format(cert))

        for b in bots_list:
            token = b.get('TOKEN', None)
            if not token:
                break

            allowed_updates = b.get('ALLOWED_UPDATES', None)
            timeout = b.get('TIMEOUT', None)
            proxy = b.get('PROXY', None)

            if self.mode == WEBHOOK_MODE:
                try:
                    if b.get('MESSAGEQUEUE_ENABLED', False):
                        q = mq.MessageQueue(all_burst_limit=b.get('MESSAGEQUEUE_ALL_BURST_LIMIT', 29),
                                            all_time_limit_ms=b.get('MESSAGEQUEUE_ALL_TIME_LIMIT_MS', 1024))
                        if proxy:
                            request = Request(proxy_url=proxy['proxy_url'], urllib3_proxy_kwargs=proxy['urllib3_proxy_kwargs'],
                                              con_pool_size=b.get('MESSAGEQUEUE_REQUEST_CON_POOL_SIZE', 8))
                        else:
                            request = Request(con_pool_size=b.get('MESSAGEQUEUE_REQUEST_CON_POOL_SIZE', 8))
                        bot = MQBot(token, request=request, mqueue=q)
                    else:
                        request = None
                        if proxy:
                            request = Request(proxy_url=proxy['proxy_url'], urllib3_proxy_kwargs=proxy['urllib3_proxy_kwargs'])
                        bot = telegram.Bot(token=token, request=request)

                    DjangoTelegramBot.dispatchers.append(Dispatcher(bot, None, workers=0))
                    hookurl = '{}/{}/{}/'.format(webhook_site, webhook_base, token)
                    max_connections = b.get('WEBHOOK_MAX_CONNECTIONS', 40)
                    setted = bot.setWebhook(hookurl, certificate=certificate, timeout=timeout, max_connections=max_connections,
                                            allowed_updates=allowed_updates)
                    webhook_info = bot.getWebhookInfo()
                    real_allowed = webhook_info.allowed_updates if webhook_info.allowed_updates else ["ALL"]

                    bot.more_info = webhook_info
                    logger.info('Telegram Bot <{}> setting webhook [ {} ] max connections:{} allowed updates:{} pending updates:{} : {}'.format(
                        bot.username, webhook_info.url, webhook_info.max_connections, real_allowed, webhook_info.pending_update_count, setted))

                except InvalidToken:
                    logger.error('Invalid Token : {}'.format(token))
                    return
                except TelegramError as er:
                    logger.error('Error : {}'.format(repr(er)))
                    return

            else:
                try:
                    updater = Updater(token=token, request_kwargs=proxy)
                    bot = updater.bot
                    bot.delete_webhook()
                    DjangoTelegramBot.updaters.append(updater)
                    DjangoTelegramBot.dispatchers.append(updater.dispatcher)
                    DjangoTelegramBot.__used_tokens.add(token)
                except InvalidToken:
                    logger.error('Invalid Token : {}'.format(token))
                    return
                except TelegramError as er:
                    logger.error('Error : {}'.format(repr(er)))
                    return

            DjangoTelegramBot.bots.append(bot)
            DjangoTelegramBot.bot_tokens.append(token)
            DjangoTelegramBot.bot_usernames.append(bot.username)

        logger.debug('Telegram Bot <{}> set as default bot'.format(DjangoTelegramBot.bots[0].username))

        def module_imported(module_name, method_name, execute):
            try:
                m = importlib.import_module(module_name)
                if execute and hasattr(m, method_name):
                    logger.debug('Run {}.{}()'.format(module_name, method_name))
                    getattr(m, method_name)()
                else:
                    logger.debug('Run {}'.format(module_name))

            except ImportError as er:
                if settings.DJANGO_TELEGRAMBOT.get('STRICT_INIT'):
                    raise er
                else:
                    logger.error('{} : {}'.format(module_name, repr(er)))
                    return False

            return True

        # import telegram bot handlers for all INSTALLED_APPS
        for app_config in apps.get_app_configs():
            if module_has_submodule(app_config.module, TELEGRAM_BOT_MODULE_NAME):
                module_name = '%s.%s' % (app_config.name, TELEGRAM_BOT_MODULE_NAME)
                if module_imported(module_name, 'main', True):
                    logger.info('Loaded {}'.format(module_name))

        num_bots=len(DjangoTelegramBot.__used_tokens)
        if self.mode == POLLING_MODE and num_bots>0:
            logger.info('Please manually start polling update for {0} bot{1}. Run command{1}:'.format(num_bots, 's' if num_bots>1 else ''))
            for token in DjangoTelegramBot.__used_tokens:
                updater = DjangoTelegramBot.get_updater(bot_id=token)
                logger.info('python manage.py botpolling --username={}'.format(updater.bot.username))
